File: examples/03-relay-realtime/ports/go/main.go
// Workshop RELAY realtime — Pillar 3 (Go).
//
// Opens a SignalWire RELAY connection in the background, forwards events
// over a gorilla/websocket browser WS, exposes outbound /api/dial.

package main

import (
	"context"
	"encoding/json"
	"log"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"sync"

	"github.com/gorilla/websocket"
	"github.com/signalwire/signalwire-go/pkg/relay"
	"github.com/signalwire/signalwire-go/pkg/rest"
)

type creds struct{ ProjectID, Space, Token string }

var (
	mu          sync.Mutex
	c           *creds
	relayClient *relay.Client
	subs        = make(map[*websocket.Conn]struct{})
)

var upgrader = websocket.Upgrader{
	CheckOrigin: func(r *http.Request) bool { return true },
}

func broadcast(event map[string]any) {
	mu.Lock()
	defer mu.Unlock()
	for ws := range subs {
		if err := ws.WriteJSON(event); err != nil {
			ws.Close()
			delete(subs, ws)
		}
	}
}

func writeJSON(w http.ResponseWriter, status int, body any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(body)
}

func sharedUIDir() string {
	wd, _ := os.Getwd()
	return filepath.Join(wd, "..", "..", "..", "..", "shared", "ui")
}

func main() {
	port := os.Getenv("PORT")
	if port == "" {
		port = "8002"
	}
	sharedDir := sharedUIDir()

	mux := http.NewServeMux()
	mux.Handle("/shared/", http.StripPrefix("/shared/", http.FileServer(http.Dir(sharedDir))))

	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/" {
			http.NotFound(w, r)
			return
		}
		http.ServeFile(w, r, filepath.Join(sharedDir, "creds-form.html"))
	})

	mux.HandleFunc("/demo.js", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/javascript")
		http.ServeFile(w, r, "demo.js")
	})

	mux.HandleFunc("/api/setup", func(w http.ResponseWriter, r *http.Request) {
		var body struct{ ProjectID, Space, Token string }
		// JSON field names from frontend are snake_case; rebind via tags below.
		var raw struct {
			ProjectID string `json:"project_id"`
			Space     string `json:"space"`
			Token     string `json:"token"`
		}
		if err := json.NewDecoder(r.Body).Decode(&raw); err != nil {
			writeJSON(w, 400, map[string]any{"ok": false, "error": "invalid json"})
			return
		}
		body.ProjectID = strings.TrimSpace(raw.ProjectID)
		body.Space = strings.TrimSpace(raw.Space)
		body.Token = strings.TrimSpace(raw.Token)
		if body.ProjectID == "" || body.Space == "" || body.Token == "" {
			writeJSON(w, 400, map[string]any{"ok": false, "error": "All fields required"})
			return
		}
		rc, err := rest.NewRestClient(body.ProjectID, body.Token, body.Space)
		if err != nil {
			writeJSON(w, 400, map[string]any{"ok": false, "error": err.Error()})
			return
		}
		if _, err := rc.PhoneNumbers.List(map[string]any{"limit": 1}); err != nil {
			writeJSON(w, 400, map[string]any{"ok": false, "error": "Credential check failed: " + err.Error()})
			return
		}

		// Tear down any prior relay client
		mu.Lock()
		if relayClient != nil {
			relayClient.Disconnect()
			relayClient = nil
		}
		mu.Unlock()

		rl := relay.NewRelayClient(
			relay.WithProject(body.ProjectID),
			relay.WithToken(body.Token),
			relay.WithSpace(body.Space),
			relay.WithContexts("workshop"),
		)
		rl.OnCall(func(call *relay.Call) {
			broadcast(map[string]any{
				"kind":    "call",
				"state":   "incoming",
				"call_id": call.CallID,
			})
			if err := call.Answer(); err != nil {
				broadcast(map[string]any{"kind": "error", "message": "answer failed: " + err.Error()})
				return
			}
			broadcast(map[string]any{"kind": "call", "state": "answered", "call_id": call.CallID})
		})

		go func() {
			if err := rl.Connect(); err != nil {
				broadcast(map[string]any{"kind": "error", "message": "RELAY connect failed: " + err.Error()})
				return
			}
			broadcast(map[string]any{"kind": "system", "message": "RELAY connected"})
		}()

		mu.Lock()
		c = &creds{ProjectID: body.ProjectID, Space: body.Space, Token: body.Token}
		relayClient = rl
		mu.Unlock()

		writeJSON(w, 200, map[string]any{"ok": true, "jwt": "session-validated", "subscriber_id": "n/a"})
	})

	mux.HandleFunc("/api/dial", func(w http.ResponseWriter, r *http.Request) {
		mu.Lock()
		rl := relayClient
		mu.Unlock()
		if rl == nil {
			writeJSON(w, 400, map[string]any{"ok": false, "error": "Run setup first"})
			return
		}
		var body struct{ From, To string }
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			writeJSON(w, 400, map[string]any{"ok": false, "error": "invalid json"})
			return
		}
		from := strings.TrimSpace(body.From)
		to := strings.TrimSpace(body.To)
		if from == "" || to == "" {
			writeJSON(w, 400, map[string]any{"ok": false, "error": "from + to required"})
			return
		}
		devices := [][]map[string]any{{{"type": "phone", "from": from, "to": to, "timeout": 30}}}
		call, err := rl.Dial(devices)
		if err != nil {
			writeJSON(w, 400, map[string]any{"ok": false, "error": err.Error()})
			return
		}
		writeJSON(w, 200, map[string]any{"ok": true, "call_id": call.CallID})
	})

	mux.HandleFunc("/ws/events", func(w http.ResponseWriter, r *http.Request) {
		ws, err := upgrader.Upgrade(w, r, nil)
		if err != nil {
			return
		}
		mu.Lock()
		subs[ws] = struct{}{}
		hasClient := relayClient != nil
		mu.Unlock()
		if hasClient {
			_ = ws.WriteJSON(map[string]any{"kind": "system", "message": "ws connected"})
		} else {
			_ = ws.WriteJSON(map[string]any{"kind": "error", "message": "Run setup first"})
		}
		// Keep the connection open; reads drain control frames + detect close.
		for {
			if _, _, err := ws.NextReader(); err != nil {
				break
			}
		}
		mu.Lock()
		delete(subs, ws)
		mu.Unlock()
		ws.Close()
	})

	addr := "0.0.0.0:" + port
	log.Printf("RELAY realtime listening on %s", addr)
	_ = context.TODO()
	if err := http.ListenAndServe(addr, mux); err != nil {
		log.Fatal(err)
	}
}

File: examples/03-relay-realtime/ports/rust/src/main.rs
//! Workshop RELAY realtime — Pillar 3 (Rust).
//!
//! Axum WS + signalwire::relay::client::Client with on_call handler
//! broadcasting events over a tokio::sync::broadcast channel.

use axum::{
    extract::ws::{Message, WebSocket, WebSocketUpgrade},
    http::StatusCode,
    response::{IntoResponse, Json},
    routing::{get, post},
    Router,
};
use serde::Deserialize;
use serde_json::{json, Value};
use signalwire::relay::client::Client as RelayClient;
use signalwire::rest::client::RestClient;
use std::{net::SocketAddr, path::PathBuf, sync::{Arc, Mutex}};
use tokio::sync::{broadcast, OnceCell};
use tower_http::services::ServeDir;

struct AppState {
    creds: Mutex<Option<(String, String, String)>>,
    relay: Mutex<Option<Arc<RelayClient>>>,
    events_tx: broadcast::Sender<Value>,
}

static STATE: OnceCell<Arc<AppState>> = OnceCell::const_new();

fn shared_ui_dir() -> PathBuf {
    let cwd = std::env::current_dir().unwrap();
    cwd.join("..").join("..").join("..").join("..").join("shared").join("ui")
        .canonicalize().unwrap_or_else(|_| cwd.join("../../../../shared/ui"))
}

#[derive(Deserialize)]
struct SetupBody { project_id: String, space: String, token: String }

#[derive(Deserialize)]
struct DialBody { from: String, to: String }

async fn setup(Json(b): Json<SetupBody>) -> impl IntoResponse {
    let (pid, sp, tk) = (b.project_id.trim(), b.space.trim(), b.token.trim());
    if pid.is_empty() || sp.is_empty() || tk.is_empty() {
        return (StatusCode::BAD_REQUEST, Json(json!({"ok": false, "error": "All fields required"})));
    }
    if let Err(e) = RestClient::new(pid, tk, sp).phone_numbers().list(json!({"limit": 1})).await {
        return (StatusCode::BAD_REQUEST, Json(json!({"ok": false, "error": format!("Credential check failed: {}", e)})));
    }

    let state = STATE.get().unwrap().clone();
    *state.creds.lock().unwrap() = Some((pid.into(), sp.into(), tk.into()));

    let tx = state.events_tx.clone();
    let rl = Arc::new(RelayClient::new(pid, tk, sp, &["workshop"]));
    let rl_clone = rl.clone();
    let tx_clone = tx.clone();

    rl.on_call(move |call| {
        let _ = tx_clone.send(json!({
            "kind": "call", "state": "incoming", "call_id": call.call_id(),
        }));
        let tx2 = tx_clone.clone();
        let call2 = call.clone();
        tokio::spawn(async move {
            match call2.answer().await {
                Ok(_) => { let _ = tx2.send(json!({"kind": "call", "state": "answered", "call_id": call2.call_id()})); }
                Err(e) => { let _ = tx2.send(json!({"kind": "error", "message": format!("answer failed: {}", e)})); }
            }
        });
    });

    let tx2 = tx.clone();
    tokio::spawn(async move {
        match rl_clone.connect().await {
            Ok(_) => { let _ = tx2.send(json!({"kind": "system", "message": "RELAY connected"})); }
            Err(e) => { let _ = tx2.send(json!({"kind": "error", "message": format!("RELAY connect failed: {}", e)})); }
        }
    });

    *state.relay.lock().unwrap() = Some(rl);
    (StatusCode::OK, Json(json!({"ok": true, "jwt": "session-validated", "subscriber_id": "n/a"})))
}

async fn dial(Json(b): Json<DialBody>) -> impl IntoResponse {
    let state = STATE.get().unwrap().clone();
    let rl = match state.relay.lock().unwrap().clone() {
        Some(r) => r,
        None => return (StatusCode::BAD_REQUEST, Json(json!({"ok": false, "error": "Run setup first"}))),
    };
    let from = b.from.trim();
    let to = b.to.trim();
    if from.is_empty() || to.is_empty() {
        return (StatusCode::BAD_REQUEST, Json(json!({"ok": false, "error": "from + to required"})));
    }
    let devices = vec![vec![json!({
        "type": "phone", "from": from, "to": to, "timeout": 30
    })]];
    match rl.dial(devices).await {
        Ok(call) => (StatusCode::OK, Json(json!({"ok": true, "call_id": call.call_id()}))),
        Err(e) => (StatusCode::BAD_REQUEST, Json(json!({"ok": false, "error": e.to_string()}))),
    }
}

async fn ws_handler(ws: WebSocketUpgrade) -> impl IntoResponse {
    ws.on_upgrade(handle_ws)
}

async fn handle_ws(mut socket: WebSocket) {
    let state = STATE.get().unwrap().clone();
    let mut rx = state.events_tx.subscribe();
    let opening = if state.relay.lock().unwrap().is_some() {
        json!({"kind": "system", "message": "ws connected"})
    } else {
        json!({"kind": "error", "message": "Run setup first"})
    };
    if socket.send(Message::Text(opening.to_string())).await.is_err() {
        return;
    }
    loop {
        tokio::select! {
            event = rx.recv() => {
                match event {
                    Ok(e) => if socket.send(Message::Text(e.to_string())).await.is_err() { return; },
                    Err(_) => return,
                }
            }
            msg = socket.recv() => {
                match msg {
                    Some(Ok(_)) => {} // ignore client messages
                    _ => return,
                }
            }
        }
    }
}

#[tokio::main]
async fn main() {
    let (events_tx, _) = broadcast::channel::<Value>(256);
    STATE.set(Arc::new(AppState {
        creds: Mutex::new(None),
        relay: Mutex::new(None),
        events_tx,
    })).ok();

    let shared = shared_ui_dir();
    let shared_for_root = shared.clone();
    let app = Router::new()
        .route("/api/setup", post(setup))
        .route("/api/dial", post(dial))
        .route("/ws/events", get(ws_handler))
        .route("/", get(move || {
            let p = shared_for_root.join("creds-form.html");
            async move { (
                [(axum::http::header::CONTENT_TYPE, "text/html")],
                std::fs::read_to_string(&p).unwrap_or_default(),
            ) }
        }))
        .route("/demo.js", get(|| async {
            (
                [(axum::http::header::CONTENT_TYPE, "application/javascript")],
                std::fs::read_to_string("demo.js").unwrap_or_default(),
            )
        }))
        .nest_service("/shared", ServeDir::new(&shared));

    let port: u16 = std::env::var("PORT").ok().and_then(|s| s.parse().ok()).unwrap_or(8002);
    let addr: SocketAddr = ([0, 0, 0, 0], port).into();
    println!("RELAY realtime listening on http://{}", addr);
    let listener = tokio::net::TcpListener::bind(addr).await.unwrap();
    axum::serve(listener, app).await.unwrap();
}
